Gear extruder sync

thisiscam opened the pull request:
Hi!

This starts a formal PR to bringing the extruder syncing feature into HH. I ended up cleaning up some of my hacky code and properly implemented a `manual_extruder_stepper` plugin that allows a stepper to sync to an extruder's motion queue, or be manually moved when it is not synced. 

Currently this code should work, and it prints well with my printer --- it seems to have solved my friction issues! However, this is not extensively tested so far, so more eyes (or print time) are needed.

Please let me know of any feedbacks. I will also keep work on this branch for any potential fixes/improvements. 

Thank you

moggieuk commented:
Cool.  I'm interested to see where this goes.  I was wondering about configuring the gear motor as an extruder and then enabling the syncing with main extruder or unsyncing if necessary.

One issue with this is that the "endstop" on the gear motor is used for sensorless selector homing.  Not intuitive but that mode of operation needs two endstops.

thisiscam commented:
>  I was wondering about configuring the gear motor as an extruder and then enabling the syncing with main extruder or unsyncing if necessary.

The added helper class basically merges a manual stepper and an extruder. I _think_ it should work anyway we'd like it, and it should support homing just like a regular manual stepper --- at least that's my intention! But the homing part is totally untested for now.

> One issue with this is that the "endstop" on the gear motor is used for sensorless selector homing. 
 
Sorry, I didn't know really understood what sensorless selector homing was about. I will definitely look into this when I get my current stable version!

bombela commented:
> It is crucial, however, to maintain precise rotational distances for both the primary extruder stepper and the gear stepper. A mismatch in filament transfer speeds between these components could lead to undue stress and filament grinding.

In my testing with synced extruder, I found it difficult to maintain a perfect rotational distances between the two. Ultimately it is better for the gear stepper to push more than the toolhead is pulling, otherwise it would totally defeat the purpose if it were to do the opposite. With fine tuning you can surely get it to grind rarely, but every skip of the gears against the filament produced a bit of filament dust for me. Dust that would slowly accumulate in the PTFE tube.

In my testing I lowered the current for the gear motor as low as I could (0.1A), but it still had enough torque to grind the filament instead of the stepper gently missing steps.

At the end I found the noise of the ERCF gear under motion unbearable. It sounds like you are trying to murder the soul of the printer or something along those line. I cannot help but think excessive noise is an indication of something wearing out prematurely.

thisiscam commented:
>  I found it difficult to maintain a perfect rotational distances between the two.

Could you elaborate on your testing procedure for how you decided that the two don't have a matching rotational distance? 
What I did was basically first calibrate the two steppers separately with the usual caliper approach. I then load the filament, take out the PTFE tube that come out of the ERCF encoder, cut the filament, then request moves to see if the cut point is always in sync.  Basically, I think I can dial it this way in to a fair accuracy, where I can extrude ~1m filament without observing and sync issues. 

Now, this amount of tuning definitely can't be perfect, but here's my thought: In a traditional bowden setup, the filament is compressed in the Bowden tube all the time. The extruder stepper simply skips or loose steps when the pressure is waaay too high, and this is basically expected behavior. I think the pressure build up in this synced setup is actually far less than the bowden case, since we actually have the toolhead stepper pulling on the filament, instead of a melt zone that creates a ton of friction. So overall, I think a tad bit of out-of-sync is actually ok --- the ERCF gear is actually going to fighting off the extra friction on the filament buffer and rolls, and loosing steps or slipping just a tiny bit once in a while shouldn't really be a concern.

What really matters is that we should make sure the toolhead looses as few steps (or slip) as possible -- and yes, to do that, we probably want the gear extruder to "push more" than "lag behind". But this should be relatively easy to tune --- if you actually observed depressed filament in your bowden tube, just decrease the rotational distance of the gear stepper bit by bit, until you don't see that.

Further, it should be clear that the additional gear stepper's pulling force is not going to decrease your toolhead extruder's performance --- that case can only happen if the filament depression in the Bowden tube is greater than all the friction in the entire Bowden path. Unless you have a super low friction path, I don't see how that can happen... 

> With fine tuning you can surely get it to grind rarely, but every skip of the gears against the filament produced a bit of filament dust for me. Dust that would slowly accumulate in the PTFE tube.

This is true for a classic bowden setup as well (it's probably even true for direct drive, where tiny filament dust will get into the very short bowden tube that goes into the hotted). I think the only way to solve this is larger gears with better grip? I think the concern here is that too much dust might increase the friction of the PTFE tubes, and that is probably a valid concern. But, I think more testing is needed to see if this is actually a real issue. Remember, with an additional ERCF gear, we are getting around twice of the pulling force, or even more! In fact, with ERCF we can even use long NEMA 17 steppers and get a super strong pull. 

> At the end I found the noise of the ERCF gear under motion unbearable. It sounds like you are trying to murder the soul of the printer or something along those line. I cannot help but think excessive noise is an indication of something wearing out prematurely.

I'm not sure where your noise might have come from --- most likely it's the not-so-perfect meshing of the bondtech gears (with current official ERCF release, the D-shaft might get bent when the servo pushes down), and the encoder gear also makes some noise. But other than that, I think my current setup's noise is not worse than my Bowden printer. 

Moreover, this PR preserves backward compatibility (at least I tried my best, more testing will be needed), so if you don't like the noise and have a low friction setup, you would simply keep your existing config and your gears will stay synced :) 
Overall, after proper testing, I think this could turn out to be a great solution to people who must have a high friction filament path setup (e.g., due to constrained space). 

Lastly, I'd like to point out that there are many printers out there that does this kind of "synced extrusion" system --- most of time because there just have to be a long or high friction filament path.  
One recent example is bambu labs AMS system. The setup is slightly different there because _I think_ their remote extruder is a DC motor instead of stepper (wow! not even a stepper is ok for syncing). What they do is this simple spring based filament buffer https://us.store.bambulab.com/products/filament-buffer to avoid some of syncing issues.

moggieuk commented:
Thank you for this contribution!  I'm going to need to find some quite time to go through in detail.  There is a lot of logic here that seems to make sense but I'd like to understand it better. 

I wonder if it is worth adding some "filament tension equalization" logic so that after a specified amount of extrusion without tool change, the servo is set to up, followed by a return to down -- this would work to rest any accumulated mismatch?

thisiscam commented:
> Thank you for this contribution! I'm going to need to find some quite time to go through in detail. There is a lot of logic here that seems to make sense but I'd like to understand it better.

Np. Let me know how I can be of help!

> I wonder if it is worth adding some "filament tension equalization" logic so that after a specified amount of extrusion without tool change, the servo is set to up, followed by a return to down -- this would work to rest any accumulated mismatch?

This is a great idea, and I have thought about implementing it. 
My idea of implementing it uses `register_timer` to periodically check for extruder position, and enqueue a short servo up-down movement every time the extruder travelled more than a configurable threshold.
But I got stuck on how to figure out the actual travelled position of the extruder. The problem is that one can reset the extruder position by gcode (and a lot of slicer does this), and that will cause the above approach to become invalid, I think.
Alternatively, we could use the encoder. But then we lost the direction and will get more frequent triggering when there is a lot of retraction.

One concern is that this might introduce artifacts due to changing stress in the filament path. But more experiment is needed to see if that matters at all... To be fair, I have been printing with the current code for a week with no problem at all. The ERCF gear is definitely grinding the filament, but I don't think that's (majorly) caused by unsynced steppers. It's just the ERCF gears bitting off filaments --- same thing happens for any extruder that keeps moving the filament.

bombela commented:
> I actually disagree on this point. I would agree that it's not ideal for the gear stepper to lag behind. But, I don't think that will be worse than not having a gear stepper. Not having a gear stepper means that the extruder stepper alone must combat all the friction.

It is holding the filament back. How can it not be possibly worse than doing nothing? Plus now that the filament is under tension, it shortens the bends in the PTFE tube, effectively increasing friction further.

If I can attempt an analogy, take a tight turn with a car with a locked differential.

> The ERCF gear is definitely grinding the filament, but I don't think that's (majorly) caused by unsynced steppers.

In all my testing, I have never seen the gears grinding the filament (in the toolhead or the ercf) unless the filament is blocked. By clog in the nozzle, a spool tangled or what not.

Pedantic note: the steppers are perfectly synced. Its the feeding ratio that is not. And I don't think we can ever get it correct. And the risk of holding back the filament forces us to bias toward pushing too much. But I guess it were to grind infrequently enough, this might well be reasonable?

In my testing, I had a gear tooth slip every 4s. How often is yours?

Note how the bambulab printers measure the tension of the filament and adjust the feeding from that. I would hope they did not add all this complexity and cost for sheer fun.

> I'm not sure where your noise might have come from --- most likely it's the not-so-perfect meshing of the bondtech gears (with current official ERCF release, the D-shaft might get bent when the servo pushes down), and the encoder gear also makes some noise. But other than that, I think my current setup's noise is not worse than my Bowden printer.

I suppose tolerance for noise is subjective. The noise of the servo is louder than my printer at full speed for example. In fact, the case the electronic fans are actually the loudest part of the printer (and they are noctua fans). And I do print quite fast at 16m/s² acceleration and 18mm³/s flow. The ERCF gear is noisy even without the servo down. Its just how it is with so many little things rattling around.

In any case, this discussion shouldn't prevent adding this feature to happy hare. It is still very useful for testing and opens the door for further exploration. It is interesting to compare or observations and experience.

thisiscam commented:
> It is holding the filament back. How can it not be possibly worse than doing nothing? 

I think we need to first agree on what metric we are comparing on, that is, what's the definition of good or worse? 
My definition is based on the extrusion force, that is, the force acted on a piece of molten plastic in the nozzle.
It should be the case that adding another stepper does not decrease the extrusion force, as long as both steppers are working at their intended range of rpms.

Note that this is not different from people linking two steppers mechanically with timing belts. Yes, the steppers can never be tuned to be perfectly in sync, but the slight out-of-syncness shouldn't really matter for our use case.

> If I can attempt an analogy, take a tight turn with a car with a locked differential.

 Sorry.. I am not familiar with car mechanics so not sure if I get the analogy here..
I tend to think of this more like a tug of war. Will adding a person, who is trained to pull mostly synchronously with the rest of your team, decrease the total pulling force of your team? 


> In all my testing, I have never seen the gears grinding the filament (in the toolhead or the ercf) unless the filament is blocked. By clog in the nozzle, a spool tangled or what not.

Did we have two definitions of "grinding"? By grinding, I just mean that the gears bite into the filament and therefore leaves off tiny dust over time. In particular, I do not intend to mean "slipping".
Take a look at this CNC kitchen video which explains how the filament gets gripped by the bondtech gears https://www.youtube.com/watch?v=5UEUFDP8M30&t=240s --- around 3:10 he explained that the filament is expected to be bited by the gears. This will definitely grind off filament particles over time. 

> Pedantic note: the steppers are perfectly synced. Its the feeding ratio that is not. And I don't think we can ever get it correct. And the risk of holding back the filament forces us to bias toward pushing too much. 

I totally agree that we can never get the feeding ratio (and that is indeed what I meant by saying "perfectly synched") completely correct --- that's just physically impossible. 

> And the risk of holding back the filament forces us to bias toward pushing too much

Yup, I definitely lean on the safe side.

> But I guess it were to grind infrequently enough, this might well be reasonable?

I think this is talking about "slip"? If so, then yes, I think ideally we would want the ERCF gears to slip very occasionally over time. The idea  @moggieuk  brought up is to basically control that "slip". 
But to be fair, I have not seen evidence that this is necessary with one week of printing. 
If filament slips on the ERCF gears occasionally, that's totally fine --- maybe we will get more dust but I don't think that matters. 
If filament slips on the extruder gears, then it should be slipping forward, since I leaned on the safe side of letting ERCF push tad bit more than the extruder. This might creates a tiny tiny bit of over-extrusion --- but either this didn't happen or the over-extrusion is way too tiny to be observable, my prints so far look the same to me without ERCF.

> In my testing, I had a gear tooth slip every 4s. How often is yours?

If you let me know how you measured the slipping frequency, that would be very appreciated? I cannot see the gears when printing, nor can I hear slippage. 
But, if it's really slipping every 4s, I highly suspect that the two steppers are not configured correctly? As I said in the calibration note in my previous comment, I made sure that the two steppers can extruder synchronously over 1 meter without any out-of-sync-ness.  So I don't think I can observe slipping at that kind of frequency.

> Note how the bambulab printers measure the tension of the filament and adjust the feeding from that. I would hope they did not add all this complexity and cost for sheer fun.

The bambu AMS uses [DC motors](https://us.store.bambulab.com/products/ams-internal-hub-motor). So the tensioning mechanism is likely just to control how fast to spin these motors (maybe with a PID or something). With steppers, we should not need this. 

> I suppose tolerance for noise is subjective. The noise of the servo is louder than my printer at full speed for example. In fact, the case the electronic fans are actually the loudest part of the printer (and they are noctua fans). And I do print quite fast at 16m/s² acceleration and 18mm³/s flow. The ERCF gear is noisy even without the servo down. Its just how it is with so many little things rattling around.

Yeah I agree noise measurement is subjective, and could really be one complaint of such an approach. But, with the correct steppers, one shouldn't really expect too much noise increase (in db), since as you pointed out, the fans are the most noisy usually. 

> In any case, this discussion shouldn't prevent adding this feature to happy hare. It is still very useful for testing and opens the door for further exploration. It is interesting to compare or observations and experience.

Yup, np at all! I just want to push this out there since this actually worked quite well for me so far. And thank you for your comments and inputs :)

thisiscam commented:
Actually thinking about sync tuning a bit more. It might be possible to tune the steppers so that they can only differ very minimally after extruding a total length of filament that fills the entire print volume.. In that Case, we do get "perfectly synchronized steppers"

moggieuk commented:
I'm taking my time thinking on this.  I am in the process of updating Happy Hare for a v2 release where, along with a number of other things I want to simply options.   One area is in the sync_load_length/speed and sync_unload_length/speed.

I want to replace all of the sync logic with three simple booleans:
sync_when_loading
sync_when_unloading
sync_when_printing

And two (extruder) speed settings:
load_speed
unload_speed

I think it makes sense to use only one mechanism for "syncing" steppers so I was thinking to always configure the gear stepper as a manual_extruder_stepper defined in this PR.  Given that Klipper has improved the erroneous Timer Too Close errors I think it is safe to make the defaults: sync on for load/unload, sync off for print.  Less setup options is much easier to support.

I do want to think more about relieving tension between extruder and gear.  I have always thought some form of feedback from strain gauge is the way to go (easy to implement but more hardware/electronics).  The servo release would be a noisy kluge.  Tuning gear stepper current another kludge.

Maybe the strain release mechanism could be either or... either you have some form of feedback switch or a "poor-mans" approach is used.

thisiscam commented:
> I'm taking my time thinking on this. I am in the process of updating Happy Hare for a v2 release where, along with a number of other things I want to simply options. One area is in the sync_load_length/speed and sync_unload_length/speed.
> 
> I want to replace all of the sync logic with three simple booleans: sync_when_loading sync_when_unloading sync_when_printing
> 
> And two (extruder) speed settings: load_speed unload_speed
> 
> I think it makes sense to use only one mechanism for "syncing" steppers so I was thinking to always configure the gear stepper as a manual_extruder_stepper defined in this PR. Given that Klipper has improved the erroneous Timer Too Close errors I think it is safe to make the defaults: sync on for load/unload, sync off for print. Less setup options is much easier to support.

Makes sense to me. I tried not to touch existing syncing logic since I don't want to break existing users. 
Given that the current PR does preserve existing functionality, I think it make sense to merge it first (onto master or perhaps to a development branch) just so that we can proceed with removing the syncing logic that doesn't use `manual_extruder_stepper`? I can do that for sure.
Otherwise, do you want me create a separate PR for the `manual_extruder_stepper` module? I think in that case we'd end up recreating much of what this PR does, TBH.

> I do want to think more about relieving tension between extruder and gear. I have always thought some form of feedback from strain gauge is the way to go (easy to implement but more hardware/electronics). The servo release would be a noisy kluge. Tuning gear stepper current another kludge.
> 
> Maybe the strain release mechanism could be either or... either you have some form of feedback switch or a "poor-mans" approach is used.

Also makes sense. My only comment is that I think we can be less conservative here --- I think a lot of ERCF users are actually having friction problems and could potentially benefit from the current PR. 
From me running this setup for two weeks, I haven't seen any issue with syncing naively. Maybe I'm too tolerate of print quality? But I honestly can't really tell any difference after enabling syncing vs without ERCF. I think it makes sense to let more people run this setup (beta?) to gather more data just so that we have enough evidence of how to proceed? Otherwise, we might end up in the land of premature optimization.

Don't get me wrong, it's totally slick to have a strain relief mechanism. I just think we could actually roll the naive thing out there before implementing that next level awesomeness (which a LOT needs to be changed at that point). 

As a side note, I came across [this design](https://www.printables.com/model/471788-dual-color-filament-buffer) of the filament strain release with a sensor the other day. This basically uses the same mechanism as on bambu AMS (I think). We could potentially use a similar design if we ended up actually implementing this piece of hardware. But, in my communication with the author of this design, I was told that this kind of strain relief is probably only going to be useful when the remote extruder is a DC motor --- this is also partly why I think we shouldn't go to the over-engineering route this fast.

> (JFYI: I recently added a "flow_rate %" variable to ercf_encoder. This gives a % mismatch of the extruder to encoder movement. Some observations I made is that because of the jerky movement of the encoder I needed to sample over 10-20 movements to get a stable result. Also, even with a perfectly calibrated and working encoder and no friction there is still some slip 1-2%. All this to say that I don't think our current encoder is going to be much help here UNLESS it becomes more sophisticated and adds a strain gauge.... hmmm)

Yes, TBH, after my experience with ERCF I actually have my doubts of the usefulness of the encoder of ERCF. 
It seems to me to be slightly over-engineered and we haven't really considered the points of failure that have been introduced into the system by having the encoder. 
For example, I think just two microswitch sensors at the ERCF and at the toolhead will be more reliable. Yes, we'd get less fancy features like slippage detection, but most of the features of the ERCF encoder can be easily recovered by adding a standard filament runout sensor somewhere between the ERCF and the toolhead... Those sensors (such as BTT filament sensors) are much more reliable --- they are commercial products at the end of the day. And they are actually cheaper compared to the cost and effort one needs to properly tune a light sensor based encoder.

moggieuk commented:
Currently merging with some changes / simplifications (like only having manual_extruder_stepper and automatically fixing in upgrade).

Could of questions came up:
1. Why did you separate the servo action from the syncing?  Why not always ensure servo is down when synced and ignore servo when un-syncing?
2. Although I haven't experimented with it yet, I'm not sure I understand the benefit / reason for syncing during standalone tip forming -- I feel it could only make things worse.  The extruder has no problem unloading (including tip forming) because it is priming the buffer rather than pulling through it.
3. Simplification: I'm removing the need to supply the extruder name - I already have it. Replacing condition with a self.sync_while_printing style boolean. Any concerns?
4. If in print syncing is enabled I'm going to automatically re-enable.  No need for the user to do it since I already intercept the RESUME. Any concerns?
5. Given #4 the difference between issuing ERCF_SYNC_GEAR_MOTOR and setting the sync_in_print variable (via ERCF_TEST_CONFIG) would be that the command can be run at any time but would be reset on next tool change.  Setting the variable would persist through the print but would only activate on next tool change.  Should the ERCF_SYNC_GEAR_MOTOR command also flip sync_in_print variable?

Thanks.

thisiscam commented:
> Currently merging with some changes / simplifications (like only having manual_extruder_stepper and automatically fixing in upgrade part of install script.
> 
> Could of questions came up:
> 
> 1. Why did you separate the servo action from the syncing?  Why not always ensure servo is down when synced and ignore servo when un-syncing? in sync_gear_to_extruder()

I prefer preserving the possibility to sync the steppers but not engage the servo. This gives the user more flexibility for debugging.

> 2. Although I haven't experimented with it yet, I'm not sure I understand the benefit / reason for syncing during standalone tip forming -- I feel it could only make things worse.  

Worse by what criterion? The way I see it is that it might help tip form --- we even have an option to boost extruder current during tip forming, this extra feature directly gives you another stepper!

> The extruder has no problem unloading (including tip forming) because it is priming the buffer rather than pulling through it.

At some point during my experiment, my toolhead was not even able to push back the filament back into the buffer. 
Though I will acknowledge that currently this feature won't work so well without a toolhead sensor, since in that case the extruder needs to be able to push back at least a small amount of filament just so that the encoder can infer filament position. 
I'd prefer leave this feature out there, but undocumented --- in case someone wants to use it in the future.

> 3. Simplification: I'm removing the need to supply the extruder name - I already have it. 

Not sure how you get ahold of the extruder name? My concern is that a user might have multiple toolhead extruders and the ERCF unit only syncs to one of the toolheads. 

> Replacing condition with a self.sync_while_printing style boolean. Any concerns?

I'd prefer a name like `sync_gear_stepper`, instead of `while_printing`, since I actually use this sync feature outside of a print (like when I load filament and do test extrusions using the web UI)

> 4. If in print syncing is enabled I'm going to automatically re-enable.  No need for the user to do it since I already intercept the RESUME. Any concerns?

When is that re-enabling going to happen? It can't be when `ERCF_UNLOCK` is called because at that point the user might not have cleared out the issue. With my current workflow, I usually invoke a tool change upon fixing the hardware issue, and that will sync the steppers automatically.  

> 5. Given # 4 the difference between issuing ERCF_SYNC_GEAR_MOTOR and setting the sync_in_print variable (via ERCF_TEST_CONFIG) would be that the command can be run at any time but would be reset on next tool change.  Setting the variable would persist through the print but would only activate on next tool change.  Should the ERCF_SYNC_GEAR_MOTOR command also flip sync_in_print variable?

I think it makes sense to just error if the boolean flag (e.g. `sync_in_print `) is set to false but the user still requested to sync the steppers. Because that case might indicate that the user is not aware of the configuration status of the printer, and we don't want to sync the steppers without user's consent.

bombela commented:
> For example, I think just two microswitch sensors at the ERCF and at the toolhead will be more reliable. Yes, we'd get less fancy features like slippage detection, but most of the features of the ERCF encoder can be easily recovered by adding a standard filament runout sensor somewhere between the ERCF and the toolhead... Those sensors (such as BTT filament sensors) are much more reliable --- they are commercial products at the end of the day. And they are actually cheaper compared to the cost and effort one needs to properly tune a light sensor based encoder.

The BTT sensor is total crap in my experience. It has way too much friction, and it is a total pain to push filament through. In fact flexible filaments are impossible. Internally it is constructed with a bearing and a mouse wheel with a computer mouse optical encoder.

The ERCF encoder on the other hand, has very little friction while having a finer resolution: less than 1mm vs 10mm for the BTT). In my experience the ERCF encoder is perfectly accurate and has been 100% reliable for me. I do not appear to have any slippage.

Clog detection is really useful. Bowden correction works really well for the occasional gear stepper misses/slippage. It is also convenient not to need an extra sensor in the toolhead.

moggieuk commented:
> > 2. Although I haven't experimented with it yet, I'm not sure I understand the benefit / reason for syncing during standalone tip forming -- I feel it could only make things worse.
> 
> Worse by what criterion? The way I see it is that it might help tip form --- we even have an option to boost extruder current during tip forming, this extra feature directly gives you another stepper!

Personally I can't see this working in practice.  The steppers are separated by 600mm-1200mm of slack filament.  Certainly the the benefit in continual printing is there but this seems like it is just shaking two ends of a soft noodle.

The reason I'm asking it that I want to release this methodically and not over complicate initially.


> > The extruder has no problem unloading (including tip forming) because it is priming the buffer rather than pulling through it.
> 
> At some point during my experiment, my toolhead was not even able to push back the filament back into the buffer. Though I will acknowledge that currently this feature won't work so well without a toolhead sensor, since in that case the extruder needs to be able to push back at least a small amount of filament just so that the encoder can infer filament position. I'd prefer leave this feature out there, but undocumented --- in case someone wants to use it in the future.

Not disagreeing to the feature, but the rollout sequencing.  I know from experience how my support and chatter there is about the simplest of changes and I need to make sure everything is covered.


> > 3. Simplification: I'm removing the need to supply the extruder name - I already have it.
> 
> Not sure how you get ahold of the extruder name? My concern is that a user might have multiple toolhead extruders and the ERCF unit only syncs to one of the toolheads.

There is no feature to change the named extruder at the moment on the fly without config change.   It is also defined on the ercf_encoder.  I can see this being an addition but it should be done for all components rather than just one.   It could move to ercf_parameters (probably nicer) but then I'd want to remove from ercf_encoder hardware config. More work to do on installer/upgrader script and doc.


> > Replacing condition with a self.sync_while_printing style boolean. Any concerns?
> 
> I'd prefer a name like `sync_gear_stepper`, instead of `while_printing`, since I actually use this sync feature outside of a print (like when I load filament and do test extrusions using the web UI)

As per earlier message, I want to create a simple enablement for syncing different phases of the ERCF control.  I.e. separate extruder loading, from printing, from unloading, ...  so this is about user enablement params of when syncing occurs in ercf.py.  They all call sync_gear_stepper.  Again, I'd like to remove the "sync load length" logic I have now.

So the command is still the same as in your PR


> > 4. If in print syncing is enabled I'm going to automatically re-enable.  No need for the user to do it since I already intercept the RESUME. Any concerns?
> 
> When is that re-enabling going to happen? It can't be when `ERCF_UNLOCK` is called because at that point the user might not have cleared out the issue. With my current workflow, I usually invoke a tool change upon fixing the hardware issue, and that will sync the steppers automatically.

There are 3 cases where printing is resumed: user RESUME, endless spool tool remapping and after a normal tool change.  The point is that I can already ensure that syncing is resumed in all these cases.  No need for the user to do anything special.

FYI.  I'm having to tweak manual_extruder_stepper.py to work with Python2... still lots using it.  I.e. super().xx doesn't work.

thisiscam commented:
> The BTT sensor is total crap in my experience. It has way too much friction, and it is a total pain to push filament through. In fact flexible filaments are impossible. Internally it is constructed with a bearing and a mouse wheel with a computer mouse optical encoder.

I agree with your pain points, but these (larger bearing wheel, and a mouse rotary encoder) are actually what makes it more reliable. The fine resolution on the ERCF encoder caused it to also sense noise. The bondtech gears on the ERCF encoder are not designed to be used without active extrusion power, and they slip without really tightening the grip (think about how hard these bondtech gears grips the filament in an extruder -- this is how they are supposed to be used). 

> The ERCF encoder on the other hand, has very little friction while having a finer resolution: less than 1mm vs 10mm for the BTT). In my experience the ERCF encoder is perfectly accurate and has been 100% reliable for me. I do not appear to have any slippage.

It appears that @moggieuk have done proper testing and confirmed that there is slippage even with a nicely tuned encoder. This is also my empirical observation.

> Clog detection is really useful. Bowden correction works really well for the occasional gear stepper misses/slippage. 

Sometimes it is the encoder that slipped (even together with the gear steppers), and you might not have noticed.

> It is also convenient not to need an extra sensor in the toolhead.

Sure. But with a toolhead sensor, the encoder can be basically replaced with a simple microswitch. That shall be much more reliable than an encoder. The only functionality I think you really want from an encoder is the clog detection. I see that this can be done as a separate project from ERCF, perhaps a mini sub-project sort --- and you'd work really hard to make it work more reliably than the BTT one. 

As a side note, I think there is possibility to use extruder stall guard in place of the toolhead sensor.

thisiscam commented:
> Not disagreeing to the feature, but the rollout sequencing. I know from experience how my support and chatter there is about the simplest of changes and I need to make sure everything is covered.

Sure, fine point. I still don't see any reason to not to put it undocumented as as private API. There might be someone (like me) who wants to use it, and would like to take the effort (just like I did) to read the code to see how to get it to work.

> There is no feature to change the named extruder at the moment on the fly without config change. It is also defined on the ercf_encoder. I can see this being an addition but it should be done for all components rather than just one. It could move to ercf_parameters (probably nicer) but then I'd want to remove from ercf_encoder hardware config. More work to do on installer/upgrader script and doc.

My concern was not about changing values on the fly. I was concerned about if I have two toolheads, one for printing PVA support and the other for printing multiple materials with ERCF, can I explicitly specify that an extruder called "MMU_extruder" be synched with the ERCF gear? AFAIK, we shouldn't rely on there exists a toolhead extruder being called "extruder". 

> > > Replacing condition with a self.sync_while_printing style boolean. Any concerns?
> > 
> > 
> > I'd prefer a name like `sync_gear_stepper`, instead of `while_printing`, since I actually use this sync feature outside of a print (like when I load filament and do test extrusions using the web UI)
> 
> As per earlier message, I want to create a simple enablement for syncing different phases of the ERCF control. I.e. separate extruder loading, from printing, from unloading, ... so this is about user enablement params of when syncing occurs in ercf.py. They all call sync_gear_stepper. Again, I'd like to remove the "sync load length" logic I have now.

My point was on the naming on that boolean flag, not on whether that flag should be implemented or not. 

> There are 3 cases where printing is resumed: user RESUME, endless spool tool remapping and after a normal tool change. The point is that I can already ensure that syncing is resumed in all these cases. No need for the user to do anything special.

That sounds great. I don't know how the user RESUME works. My suggestion is just don't immediately re-sync if user invokes `ERCF_UNLOCK` --- that will cause unwanted consequences.  Endless spool tool remapping and normal tool change should already reengage the syncing, I think.

> FYI. I'm having to tweak manual_extruder_stepper.py to work with Python2... still lots using it. I.e. super().xx doesn't work.

Thanks. I haven't seen Python 2 for a while now and totally forgot about it.

bombela commented:
> It appears that @moggieuk have done proper testing and confirmed that there is slippage even with a nicely tuned encoder. This is also my empirical observation.

I get 0.44mm (less than 0.1%) of slippage between the encoder and the gear motor over a sequence of 36 load/unload of 500mm. I am curious what you get with the BTT filament motion sensor.

thisiscam commented:
> > It appears that @moggieuk have done proper testing and confirmed that there is slippage even with a nicely tuned encoder. This is also my empirical observation.
> 
> I get 0.44mm (less than 0.1%) of slippage between the encoder and the gear motor over a sequence of 36 load/unload of 500mm. I am curious what you get with the BTT filament motion sensor.

You probably have a very low friction setup. My bowden tube alone is 1000mm… not to mention the spool rolls.

If ERCF is designed the way I described, it wouldn't even need to measure slippage. The filament runout sensor (btt or not, you can even use your favorite ercf encoder if you want) is simply there to do its sole job of filament runout detection. 

This issue is getting derailed so much. We should make these discussions on discord dev channel.

bombela commented:
I load/unload at 300mm/s and 400mm/s². And print at 7.5mm/s (18mm³/s) max. I have 1.2m from toolhead to ERCF, and 0.7m from ERCF to spools. A total of 1.9m. I have not measured the friction, but I intend to acquire a suitable measuring device to do so at some point.

Discord is private and non-searchable. Here at least this is public for everybody to read and share (for now). But you are right, we are off topic. I will find you over there.

moggieuk commented:
@thisiscam Hi, I've got this mostly integrated but I get this error for the case where ManualExtruderStepper is used for gear stepper but it use as a normal ManualStepper only.  The easiest way to reproduce is simply to make sync_gear_to_extruder never call gear_stepper.sync_to_extruder(...)

def _sync_gear_to_extruder(self, sync=True):
    return
    <old logic>
    
 Any ideas? I looks as though it only works if sync_to_extruder(None) is called frequently...

Internal error on command:"ERCF_EJECT"
Internal Error on WebRequest: gcode/script
Traceback (most recent call last):
  File "/home/pi/klipper/klippy/webhooks.py", line 256, in _process_request
    func(web_request)
  File "/home/pi/klipper/klippy/webhooks.py", line 431, in _handle_script
    self.gcode.run_script(web_request.get_str('script'))
  File "/home/pi/klipper/klippy/gcode.py", line 216, in run_script
    self._process_commands(script.split('\n'), need_ack=False)
  File "/home/pi/klipper/klippy/gcode.py", line 198, in _process_commands
    handler(gcmd)
  File "/home/pi/klipper/klippy/gcode.py", line 135, in <lambda>
    func = lambda params: origfunc(self._get_extended_params(params))
  File "/home/pi/klipper/klippy/extras/ercf.py", line 2845, in cmd_ERCF_EJECT
    self._unload_tool()
  File "/home/pi/klipper/klippy/extras/ercf.py", line 2164, in _unload_tool
    self._unload_sequence(self._get_calibration_ref(), skip_tip=skip_tip)
  File "/home/pi/klipper/klippy/extras/ercf.py", line 2224, in _unload_sequence
    self._unload_encoder(length) # Full slow unload
  File "/home/pi/klipper/klippy/extras/ercf.py", line 2423, in _unload_encoder
    delta = self._trace_filament_move(msg, -self.encoder_move_step_size)
  File "/home/pi/klipper/klippy/extras/ercf.py", line 1804, in _trace_filament_move
    self._gear_stepper_move_wait(distance, accel=accel)
  File "/home/pi/klipper/klippy/extras/ercf.py", line 1775, in _gear_stepper_move_wait
    self.gear_stepper.do_move(dist, speed, accel, sync)
  File "/home/pi/klipper/klippy/extras/manual_extruder_stepper.py", line 91, in do_move
    return super(ManualExtruderStepper, self).do_move(movepos, speed, accel, sync)
  File "/home/pi/klipper/klippy/extras/manual_stepper.py", line 69, in do_move
    self.rail.generate_steps(self.next_cmd_time)
  File "/home/pi/klipper/klippy/stepper.py", line 226, in generate_steps
    raise error("Internal error in stepcompress")
error: Internal error in stepcompress

thisiscam commented:
Could you push your code? That will be easier for me to debug

thisiscam commented:
Actually,.i think I know where the problem is. Unfortunately, My printer is in maintainence so i won't be able to test it until maybe Friday. I can try to push a fix tomorrow but both guarantee that it will worm

thisiscam commented:
@moggieuk I just pushed a fix. I have not tried running it at all. But I think it should be ok.

thisiscam commented:
@moggieuk One thing that you might want to be careful with when using this manual_extruder_stepper for the sync load/unload logic: you probably want to temporarily disable PA because this syncing the gear this way will actually inherit the PA of the extruder, I think. Not sure if will be an issue but I'd imagine it to be noisier process for the encoder with PA enabled. Also disabling PA for any move that doesn't actually extrude plastic makes sense IMO.

moggieuk commented:
Maybe a reset PA when syncing to None. So regular non synced moved don’t have that artifact.

I’ll test a bit later and if I have time can push a branch to finalize things in (might not be able
To get to it until tomorrow). Basically it’s what you have in the PR with a couple of minor changes.. pause / resume / cmd line use

thisiscam commented:
@moggieuk You should be able to make your changes and push to my branch directly. 

<img width="653" alt="Screenshot 2023-06-06 at 8 26 49 AM" src="https://github.com/moggieuk/ERCF-Software-V3/assets/7684342/b27a496e-f7ae-4989-add9-172845a52784">

thisiscam commented:
> It contains:
>
> 1. Rework of TMC driver support
> 2. Your PR
> 3. Changes to your PR similar to what we have discussed.  I have four parameters. You would want them all the '1' to enable the logic you had before.  Also added support for EndlessSpool and automatic resyncing on RESUME.
> 4. Changed the form_tip filament detection logic (don't know why I wasn't leveraging toolhead sensor before)
> 5. install.sh has upgrade path to upgrade user's ercf_hardware.cfg to new manual_extruder_stepper
> 6. Other tweaks to handle corner cases and more automatic servo movement
> 7. Integrated your doc into README as feature.
>    Testing:  The happy paths seem to work well.  There are a lot of possible  failure cases I want to be sure about. Also things like bypass support and just extruder loading and unloading (like in the bypass case). Also the filament position recovery logic.
> 

All looks good to me! I'd happy to stare at the code more once it's here.
